tests/sentry/options/test_update_channels.py

```diff
@@ -0,0 +1,249 @@
+        id="Credentials cannot be updated by admin",
+    ),
+    pytest.param(
+        "manager",
+        DEFAULT_FLAGS,
+        None,
+        UpdateChannel.AUTOMATOR,
+        NotWritableReason.CHANNEL_NOT_ALLOWED,
+        NotWritableReason.CHANNEL_NOT_ALLOWED,
+        id="Default option cannot be updated by automator",
+    ),
+    pytest.param(
+        "manager",
+        FLAG_CREDENTIAL,
+        None,
+        UpdateChannel.AUTOMATOR,
+        NotWritableReason.CHANNEL_NOT_ALLOWED,
+        NotWritableReason.CHANNEL_NOT_ALLOWED,
+        id="Credentials cannot be updated by automator",
+    ),
+    pytest.param(
+        "manager",
+        FLAG_AUTOMATOR_MODIFIABLE,
+        UpdateChannel.AUTOMATOR,
+        UpdateChannel.AUTOMATOR,
+        None,
+        None,
+        id="Basic options fully owned by automator. Can update",
+    ),
+    pytest.param(
+        "manager",
+        FLAG_AUTOMATOR_MODIFIABLE,
+        UpdateChannel.CLI,
+        UpdateChannel.AUTOMATOR,
+        None,
+        NotWritableReason.DRIFTED,
+        id="Basic options fully owned by automator. Can update",
+    ),
+    pytest.param(
+        "manager",
+        FLAG_AUTOMATOR_MODIFIABLE,
+        UpdateChannel.AUTOMATOR,
+        UpdateChannel.CLI,
+        None,
+        None,
+        id="Basic options set by automator and reset by CLI",
+    ),
+]
+
+
+@pytest.fixture()
+def manager():
+    """
+    Initializes an options storage, an options cache and an options manager
+    """
+
+    c = LocMemCache("test", {})
+    c.clear()
+    store = OptionsStore(cache=c)
+    manager = OptionsManager(store=store)
+
+    default_options = settings.SENTRY_DEFAULT_OPTIONS.copy()
+    settings.SENTRY_DEFAULT_OPTIONS = {}
+    store.flush_local_cache()
+
+    yield manager
+
+    settings.SENTRY_DEFAULT_OPTIONS = default_options
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize(
+    "manager_fixture, options_flags, set_channel, check_channel, set_to_same_value_reason, reset_reason",
+    TEST_CASES,
+)
+def test_can_update(
+    manager_fixture,
+    options_flags: int,
+    set_channel: Optional[UpdateChannel],
+    check_channel: UpdateChannel,
+    set_to_same_value_reason: Optional[NotWritableReason],
+    reset_reason: Optional[NotWritableReason],
+    request: Any,
+) -> None:
+    """
+    Tests a multi-step update workflow for two options:
+    1) check that the provided channel can update the option after registration
+    2) updates the option with another channel
+    3) check whether the original channel can update it to the same value
+    4) check whether the original channel can update it to a different value
+
+    @param set_channel: the channel that sets the option first
+    @param check_channel: the channel we call can_update with
+    @param set_to_same_value_reason: The NotWritableReason we expect at point 3
+    above.
+    @param reset_reason: The NotWritableReason we expect at point 4 above.
+    """
```

Comment:
I would say this test case is too complex.  Can we simplify by breaking this into multiple test cases?  E.g. a test case where set_channel is `None` and one where its `not None`?

If you can think of more parameters where this polymorphism can be explicitly broken up I think it would be for the best.  1). I think there are some states that have not been fully covered and 2). this would help us see which states more easily.

Given the number of parameters its hard for me to tell if this coverage is complete or not.

Reply:
The test has been redesigned.